### DISCUSSION TITLE
How to select the source of displa

### DISCUSSION THREAD

**Discussion by sprig** [emacs-helm/helm#2478] [Q&A] [answered]
Hi!

I have the following expanded `helm-mini` (I know that I can also rearrange sources for the standard `helm-mini`, but this is not what I need):
``` elisp
(defun kk/helm-mini ()
  "Preconfigured `helm' lightweight version (buffer -> recentf -> bookmarks)."
  (interactive)
  (unless helm-source-buffers-list
    (setq helm-source-buffers-list
          (helm-make-source "Buffers" 'helm-source-buffers)))
  (let ((helm-ff-transformer-show-only-basename nil))
    (helm-other-buffer '(helm-source-buffers-list
                         helm-source-buffer-not-found
                         helm-source-recentf
                         helm-source-git
                         helm-source-files-in-current-dir
                         helm-source-bookmarks
                         helm-source-f-here
                         helm-source-f-home)
                       "*helm mini*")))
```

Pretty simple stuff. `helm-source-f-*` is basically like `find <DIR>|grep helm-pattern` for the appropriate dir.
`helm-source-git` is similar, except it works for git-committed files. Definition of `helm-source-git`:
``` elisp
(setq helm-source-git
  (helm-build-async-source "Files in git"
    :candidates-process
    (lambda nil (if (and (executable-find "git") (eq (call-process "git" nil nil "*kk/git-root*" "rev-parse" "--show-toplevel") 0))
                    (start-process "git ls-files" nil "bash" "-c" (string-join `("git ls-files --full-name|grep '" ,helm-pattern "'")))
                  (start-process "test" nil "test")))
    :action '(lambda (candidate) (find-file (string-join `(,(shell-command-to-string "echo -n $(git rev-parse --show-toplevel)") "/" ,candidate))))))
``` 
I'm happy with how each source works separately from the others. My problem is that `helm` reorders the sources such that candidates from `helm-source-files-in-current-dir` show above `helm-source-git`, and I do not understand why.

What am I missing?

**Comment by sprig** [comment]
Follow up:

Setting up 
```
(setq helm-mini-default-sources '(helm-source-buffers-list
                                  helm-source-buffer-not-found
                                  helm-source-recentf
                                  helm-source-git
                                  helm-source-files-in-current-dir
                                  helm-source-recent-dir-files
                                  helm-source-bookmarks
                                  helm-source-f-here
                                  helm-source-f-home))
```
and running `helm-mini` just in case there's something funky with my definition, drops the `git` source all the way to the bottom.

**Comment by thierryvolpiatto** [comment, marked as answer]
sprig ***@***.***> writes:

> I'm happy with how each source works separately from the others. My
> problem is that helm reorders the sources such that candidates from
> helm-source-files-in-current-dir show above helm-source-git, and I do
> not understand why.
>
> What am I missing?

Async sources when called go in background (just like any async process
started with start-process) and the next source is called, if this
source is synchronous, Emacs is waiting it to finish and display it,
when all sync sources are done the async one popup, if there is more
than 1 async source, the faster one wins.

PS: Have a look at helm-ls-git package.

-- 
Thierry

**Reply by sprig** [reply]
Not sure that explanation makes total sense to me, considering that some of the sources that I have usually never exit while producing candidates (e.g. `helm-source-f-home`, is essentially running `find` in my homedir). However, it is plausible enough for me to stop spending more time on this, for now. I've settled on making a sync source for now - most of my repos aren't terribly huge.
Fairly ugly but it works:
```
  (setq helm-source-git
    (helm-build-sync-source "Files in git"
      :candidates
      '(lambda nil (if (and (executable-find "git") (eq (call-process "git" nil nil "*kk/git-root*" "rev-parse" "--show-toplevel") 0))
                      (split-string-shell-command (shell-command-to-string (string-join `("git ls-files --full-name|grep '" ,helm-pattern "'"))))
                     nil))

      :action '(lambda (candidate) (find-file (string-join `(,(shell-command-to-string "echo -n $(git rev-parse --show-toplevel)") "/" ,candidate))))))
```

I'll check out your package, thanks!
